File: implementations/python/src/esque/elasticsearch/operations.py
from __future__ import annotations

import logging
from datetime import UTC, datetime
from typing import Any, cast
from urllib.parse import urlencode

from elasticsearch import ConflictError, Elasticsearch

from esque.elasticsearch.documents import (
    INDEX_DEFINITION,
    LOCK_ID_PREFIX,
    MIGRATION_INDEX,
    MigrationLock,
    MigrationRecord,
)
from esque.migration.model import MigrationFile, MigrationFileRequestDefinition

log = logging.getLogger(__name__)


class RestClientOperations:
    def __init__(self, client: Elasticsearch, migration_key: str) -> None:
        self._client = client
        self._migration_key = migration_key

    def close(self) -> None:
        self._client.close()

    def check_migration_index_exists(self) -> bool:
        log.info("Checking if migration index [%s] exists", MIGRATION_INDEX)
        exists = bool(self._client.indices.exists(index=MIGRATION_INDEX))
        log.info("Migration index [%s] %s", MIGRATION_INDEX, "exists" if exists else "does not exist")
        return exists

    def create_migration_index(self) -> None:
        log.info("Creating migration index [%s]", MIGRATION_INDEX)
        try:
            self._client.indices.create(
                index=MIGRATION_INDEX,
                settings=cast(dict[str, Any], INDEX_DEFINITION["settings"]),
                mappings=cast(dict[str, Any], INDEX_DEFINITION["mappings"]),
            )
            log.info("Migration index [%s] created", MIGRATION_INDEX)
        except ConflictError:
            log.info("Migration index [%s] already exists — another process created it first", MIGRATION_INDEX)

    def create_lock_record(self) -> None:
        log.info("Creating lock document for migration key [%s]", self._migration_key)
        lock_id = f"{LOCK_ID_PREFIX}:{self._migration_key}"
        self._client.index(
            index=MIGRATION_INDEX,
            id=lock_id,
            document=MigrationLock(date=datetime.now(UTC)).to_document(),
            op_type="create",
        )
        log.info("Lock document for migration key [%s] created", self._migration_key)

    def delete_lock_record(self) -> None:
        log.info("Deleting lock document for migration key [%s]", self._migration_key)
        lock_id = f"{LOCK_ID_PREFIX}:{self._migration_key}"
        self._client.delete(index=MIGRATION_INDEX, id=lock_id)
        log.info("Lock document for migration key [%s] deleted", self._migration_key)

    def get_migration_records(self) -> list[MigrationRecord]:
        log.info("Getting migration records for migration key [%s]", self._migration_key)
        resp = self._client.search(
            index=MIGRATION_INDEX,
            query={"bool": {"filter": [{"term": {"migration.migrationKey": self._migration_key}}]}},
            size=10000,
        )
        hits = cast(list[dict[str, Any]], resp["hits"]["hits"])
        records = [MigrationRecord.from_document(hit["_source"]) for hit in hits]
        records.sort(key=lambda r: r.order)
        log.info("Found [%d] migration records", len(records))
        return records

    def get_migration_record_for_migration_file(self, file: MigrationFile) -> MigrationRecord | None:
        log.info(
            "Getting migration record for file [%s] and migration key [%s]",
            file.metadata.filename,
            self._migration_key,
        )
        resp = self._client.search(
            index=MIGRATION_INDEX,
            query={
                "bool": {
                    "filter": [
                        {"term": {"migration.migrationKey": self._migration_key}},
                        {"term": {"migration.filename": file.metadata.filename}},
                    ]
                }
            },
        )
        hits = cast(list[dict[str, Any]], resp["hits"]["hits"])
        if len(hits) > 1:
            raise RuntimeError(
                f"found more than one migration record for file [{file.metadata.filename}]"
                f" and migration key [{self._migration_key}]"
            )
        if len(hits) == 1:
            log.info("Found existing migration record for file [%s]", file.metadata.filename)
            return MigrationRecord.from_document(hits[0]["_source"])
        log.info("No existing migration record found for file [%s]", file.metadata.filename)
        return None

    def execute_migration_definition(self, definition: MigrationFileRequestDefinition) -> None:
        log.info("Executing migration query definition")
        target = definition.path
        if definition.params:
            target += "?" + urlencode(definition.params)
        headers: dict[str, str] = {}
        if definition.content_type:
            headers["content-type"] = definition.content_type
        body: bytes | None = definition.body.encode("utf-8") if definition.body else None
        self._client.perform_request(
            definition.method,
            target,
            headers=headers,
            body=body,
        )
        log.info("Migration query definition executed successfully")

    def create_migration_record(self, record: MigrationRecord) -> None:
        if record.migration_key != self._migration_key:
            raise ValueError("migration record migration key must match operational migration key")
        log.info("Creating migration record for file [%s]", record.filename)
        self._client.index(
            index=MIGRATION_INDEX,
            document=record.to_document(),
            refresh=True,  # type: ignore[arg-type]
        )
        log.info("Migration record for file [%s] created", record.filename)

File: implementations/typescript/src/elasticsearch/operations.ts
import type { Client } from "@elastic/elasticsearch";
import { errors } from "@elastic/elasticsearch";
import type { MigrationFile, MigrationFileRequestDefinition } from "../migration/model.js";
import type { MigrationRecord } from "./documents.js";
import {
  INDEX_DEFINITION,
  LOCK_ID_PREFIX,
  MIGRATION_INDEX,
  migrationLockToDocument,
  migrationRecordFromDocument,
  migrationRecordToDocument,
} from "./documents.js";

export class RestClientOperations {
  private readonly client: Client;
  private readonly migrationKey: string;

  constructor(client: Client, migrationKey: string) {
    this.client = client;
    this.migrationKey = migrationKey;
  }

  async close(): Promise<void> {
    await this.client.close();
  }

  async checkMigrationIndexExists(): Promise<boolean> {
    return await this.client.indices.exists({ index: MIGRATION_INDEX });
  }

  async createMigrationIndex(): Promise<void> {
    try {
      await this.client.indices.create({
        index: MIGRATION_INDEX,
        settings: INDEX_DEFINITION.settings,
        mappings: INDEX_DEFINITION.mappings,
      });
    } catch (error) {
      if (isAlreadyExistsError(error)) {
        return; // another process created it first — safe
      }
      throw error;
    }
  }

  async createLockRecord(): Promise<void> {
    await this.client.index({
      index: MIGRATION_INDEX,
      id: `${LOCK_ID_PREFIX}:${this.migrationKey}`,
      document: migrationLockToDocument(new Date()),
      op_type: "create",
    });
  }

  async deleteLockRecord(): Promise<void> {
    await this.client.delete({
      index: MIGRATION_INDEX,
      id: `${LOCK_ID_PREFIX}:${this.migrationKey}`,
    });
  }

  async getMigrationRecords(): Promise<MigrationRecord[]> {
    const response = await this.client.search<Record<string, unknown>>({
      index: MIGRATION_INDEX,
      query: { bool: { filter: [{ term: { "migration.migrationKey": this.migrationKey } }] } },
      size: 10000,
    });
    const records = response.hits.hits.map((hit) =>
      migrationRecordFromDocument(hit._source as Record<string, unknown>),
    );
    records.sort((a, b) => a.order - b.order);
    return records;
  }

  async getMigrationRecordForMigrationFile(file: MigrationFile): Promise<MigrationRecord | null> {
    const response = await this.client.search<Record<string, unknown>>({
      index: MIGRATION_INDEX,
      query: {
        bool: {
          filter: [
            { term: { "migration.migrationKey": this.migrationKey } },
            { term: { "migration.filename": file.metadata.filename } },
          ],
        },
      },
    });
    const hits = response.hits.hits;
    if (hits.length > 1) {
      throw new Error(
        `found more than one migration record for file [${file.metadata.filename}] and migration key [${this.migrationKey}]`,
      );
    }
    const first = hits[0];
    if (first !== undefined) {
      return migrationRecordFromDocument(first._source as Record<string, unknown>);
    }
    return null;
  }

  async executeMigrationDefinition(definition: MigrationFileRequestDefinition): Promise<void> {
    const headers: Record<string, string> = {};
    if (definition.contentType !== null) {
      headers["content-type"] = definition.contentType;
    }
    await this.client.transport.request(
      {
        method: definition.method,
        path: definition.path,
        querystring: definition.params ?? undefined,
        body: definition.body ?? undefined,
      },
      { headers },
    );
  }

  async createMigrationRecord(record: MigrationRecord): Promise<void> {
    if (record.migrationKey !== this.migrationKey) {
      throw new Error("migration record migration key must match operational migration key");
    }
    await this.client.index({
      index: MIGRATION_INDEX,
      document: migrationRecordToDocument(record),
      refresh: true, // without it, reads immediately after won't see the record
    });
  }
}

function isAlreadyExistsError(error: unknown): boolean {
  if (!(error instanceof errors.ResponseError)) {
    return false;
  }
  const body = error.body as { error?: { type?: string } } | undefined;
  return body?.error?.type === "resource_already_exists_exception";
}
